pino.js

```diff
@@ -175,33 +220,55 @@ function asErrValue (err) {
   }
 }
 
-function genLog (level) {
-  return function (a, b, c, d, e, f, g, h, i, j, k) {
-    var base = 0
-    var obj = null
-    var params = null
-    var msg
-    var len
+function genLog (z) {
+  return function LOG (a, b, c, d, e, f, g, h, i, j, k) {
+    var l = 0
+    var m = null
+    var n = null
+    var o
+    var p
     if (typeof a === 'object' && a !== null) {
-      obj = a
-      params = [b, c, d, e, f, g, h, i, j, k]
-      base = 1
-
-      if (obj.method && obj.headers && obj.socket) {
-        obj = mapHttpRequest(obj)
-      } else if (obj.statusCode) {
-        obj = mapHttpResponse(obj)
+      m = a
+      n = [b, c, d, e, f, g, h, i, j, k]
+      l = 1
+
+      if (m.method && m.headers && m.socket) {
+        m = mapHttpRequest(m)
+      } else if (m.statusCode) {
+        m = mapHttpResponse(m)
       }
     } else {
-      params = [a, b, c, d, e, f, g, h, i, j, k]
+      n = [a, b, c, d, e, f, g, h, i, j, k]
+    }
+    p = n.length = arguments.length - l
+    if (p > 1) {
+      o = format(n, this.formatOpts)
+    } else if (p) {
+      o = n[0]
+    }
+    this.write(m, o, z)
+  }
+}
+
+function onExit (fn) {
+  var oneFn = once(fn)
+  process.on('beforeExit', handle('beforeExit'))
+  process.on('exit', handle('exit'))
+  process.on('uncaughtException', handle('uncaughtException', 1))
+  process.on('SIGHUP', handle('SIGHUP', 129))
+  process.on('SIGINT', handle('SIGINT', 130))
+  process.on('SIGQUIT', handle('SIGQUIT', 131))
+  process.on('SIGTERM', handle('SIGTERM', 143))
+  function handle (evt, code) {
+    onExit.passCode = function (code) {
+      if (oneFn.value) { oneFn = once(fn) }
+      oneFn(code, evt)
     }
-    len = params.length = arguments.length - base
-    if (len > 1) {
-      msg = format(params, null)
-    } else if (len) {
-      msg = params[0]
+    onExit.insertCode = function () {
+      if (oneFn.value) { oneFn = once(fn) }
+      oneFn(code, evt)
     }
-    this.stream.write(this.asJson(obj, msg, level))
+    return (code === undefined) ? onExit.passCode : onExit.insertCode
```

Comment:
Are we sure there is no better way to do this?

Maybe we can expose it to the user, something like `logger.flush()`, and a snippet module to do things like

```js
require('on-exit')(logger.flush.bind(logger))
```

And an option `flushOnExit` option?